### ISSUE TITLE
Does not build on D 2.089+

### ISSUE THREAD

**Issue by p-mitana** [vnayar/s2geometry-d#4]
Library does not build with the latest DMD. I've tested different versions and it fails on 2.089 and above.

```
s2geometry-d ~master: building configuration "library"...
/usr/include/dmd/phobos/std/math.d(606,20): Error: `pure` function `std.math.abs!(const(ExactFloat)).abs` cannot call impure function `s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp`
/usr/include/dmd/phobos/std/math.d(606,20): Error: `@nogc` function `std.math.abs!(const(ExactFloat)).abs` cannot call non-@nogc function `s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp`
/usr/include/dmd/phobos/std/math.d(606,33): Error: `pure` function `std.math.abs!(const(ExactFloat)).abs` cannot call impure function `s2.util.math.exactfloat.ExactFloat.opUnary!"-".opUnary`
/usr/include/dmd/phobos/std/math.d(606,33): Error: `@nogc` function `std.math.abs!(const(ExactFloat)).abs` cannot call non-@nogc function `s2.util.math.exactfloat.ExactFloat.opUnary!"-".opUnary`
/usr/include/dmd/phobos/std/math.d(606,20): Error: function `s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp` is not `nothrow`
/usr/include/dmd/phobos/std/math.d(595,6): Error: `nothrow` function `std.math.abs!(const(ExactFloat)).abs` may throw
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(320,43): Error: template instance `std.math.abs!(const(ExactFloat))` error instantiating
/usr/include/dmd/phobos/std/algorithm/iteration.d(604,19):        instantiated from here: `__lambda1!(const(ExactFloat))`
/usr/include/dmd/phobos/std/algorithm/iteration.d(499,16):        instantiated from here: `MapResult!(__lambda1, const(ExactFloat)[])`
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(320,46):        instantiated from here: `map!(const(ExactFloat)[])`
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(442,19):        instantiated from here: `Vector!(ExactFloat, 3LU)`
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/s2predicates.d(1492,20):        instantiated from here: `Vector3!(ExactFloat)`
```

**Comment by vnayar** [comment]
I used the "dvm" tool to find that the first version in which this compilation error appears is 2.089.0, just as you reported.  I'll have a fix in a day or so as time permits.  I'm curious to find out what the cause is.

**Comment by vnayar** [comment]
Looks like `abs()` no longer works on types that have throwable functions:  https://github.com/dlang/phobos/commit/56c1473f2acdeecde249d8ec731781b0683164b3

I'm converting everything in ExactFloat to be `nothrow`.

**Comment by vnayar** [comment]
Fixed in latest commit, will cut a patch-release soon.

**Comment by vnayar** [comment]
Fix released in v0.2.3.

**Comment by the-Mitu** [comment]
Seems like the solution is not complete. `nothrow` is sorted out, but the compiler still complains about `@nogc` and `pure`.  

```
/usr/include/dmd/phobos/std/math.d(606,20): Error: pure function std.math.abs!(const(ExactFloat)).abs cannot call impure function s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp
/usr/include/dmd/phobos/std/math.d(606,20): Error: @nogc function std.math.abs!(const(ExactFloat)).abs cannot call non-@nogc function s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp
/usr/include/dmd/phobos/std/math.d(606,33): Error: pure function std.math.abs!(const(ExactFloat)).abs cannot call impure function s2.util.math.exactfloat.ExactFloat.opUnary!"-".opUnary
/usr/include/dmd/phobos/std/math.d(606,33): Error: @nogc function std.math.abs!(const(ExactFloat)).abs cannot call non-@nogc function s2.util.math.exactfloat.ExactFloat.opUnary!"-".opUnary
../../.dub/packages/s2geometry-d-0.2.3/s2geometry-d/src/s2/util/math/vector.d(320,43): Error: template instance std.math.abs!(const(ExactFloat)) error instantiating
/usr/include/dmd/phobos/std/algorithm/iteration.d(604,19):        instantiated from here: __lambda1!(const(ExactFloat))
/usr/include/dmd/phobos/std/algorithm/iteration.d(499,16):        instantiated from here: MapResult!(__lambda1, const(ExactFloat)[])
../../.dub/packages/s2geometry-d-0.2.3/s2geometry-d/src/s2/util/math/vector.d(320,46):        instantiated from here: map!(const(ExactFloat)[])
../../.dub/packages/s2geometry-d-0.2.3/s2geometry-d/src/s2/util/math/vector.d(442,19):        instantiated from here: Vector!(ExactFloat, 3LU)
../../.dub/packages/s2geometry-d-0.2.3/s2geometry-d/src/s2/s2predicates.d(1492,20):        instantiated from here: Vector3!(ExactFloat)
/usr/bin/dmd failed with exit code 1.
```

**Comment by vnayar** [comment]
Oh interesting, I wasn't seeing that error locally.  I'll give it a go after work tonight and resolve that as well.

**Comment by vnayar** [comment]
Give it another try on your end from `master`.  I have to say that the fallout of these attributes went pretty far into the code.  I'm not sure what lesson I should be learning from this experience.

**Comment by the-Mitu** [comment]
Almost...

```
/usr/include/dmd/phobos/std/math.d(606,20): Error: @nogc function std.math.abs!(const(ExactFloat)).abs cannot call non-@nogc function s2.util.math.exactfloat.ExactFloat.opCmp!int.opCmp
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(320,43): Error: template instance std.math.abs!(const(ExactFloat)) error instantiating
/usr/include/dmd/phobos/std/algorithm/iteration.d(604,19):        instantiated from here: __lambda1!(const(ExactFloat))
/usr/include/dmd/phobos/std/algorithm/iteration.d(499,16):        instantiated from here: MapResult!(__lambda1, const(ExactFloat)[])
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(320,46):        instantiated from here: map!(const(ExactFloat)[])
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/util/math/vector.d(442,19):        instantiated from here: Vector!(ExactFloat, 3LU)
../../.dub/packages/s2geometry-d-master/s2geometry-d/src/s2/s2predicates.d(1492,20):        instantiated from here: Vector3!(ExactFloat)

```

**Comment by vnayar** [comment]
This whole experience has been quite informative all around.  I also
discovered that without doing a `dub clean`, sometimes the code would
compile without any error and I would assume the problem was resolved.  But
now I did both `dub clean` before `dub build` and hopefully it's fixed
now.  It would be nice if the compiler could infer whether a function is
pure or `@nogc` on its own without it having to be explicitly spelled out
for every function.

Give it another go and let me know if it builds on your side.

**Comment by the-Mitu** [comment]
I removed the package and did `dub fetch s2geometry-d@~master`, then `dub build`. Failed, still the same error.

**Comment by vnayar** [comment]
Ok, try once more :P 

I discovered that UFCS doesn't actually work if the module which defines the function isn't available wasn't declared in the template, and if you do declare it in the template, it creates a conflict with the any generic typed function with the same signature.

Specifically, the `abs()` function in exactfloat.d wasn't actually being called at all, it was calling the one from `std.math`.  If I tried importing `exactfloat.d` from `vector.d`, the definition of `abs()` would clash would the one from `std.math`.  The whole issue about `@nogc` was a bit of a red herring, because the real problem was that it was calling the wrong function.

The solution was to add an exactfloat member function for `abs()`, so instead of `a.abs()` being interpreted as UFCS for `std.math : abs`, it actually went to the function in `ExactFloat`.
